android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentService.java

```diff
@@ -0,0 +1,876 @@
+                    default:
+                        super.handleMessage(msg);
+                }
+            } catch (Throwable e) {
+                Log.e(LOG_TAG, "Exception " + Log.getStackTraceString(e));
+            }
+        }
+
+        private boolean isValidObserver(Messenger observer) {
+            return lifecycleObservers.contains(observer);
+        }
+    }
+
+    private class GreengrassOutputStream extends OutputStream {
+
+        private final int messageId;
+        private final String extraId;
+
+        private final StringBuffer buffer = new StringBuffer(256);
+
+        GreengrassOutputStream(int messageId, String extraId) {
+            super();
+            this.messageId = messageId;
+            this.extraId = extraId;
+        }
+
+        @Override
+        public void write(int i) throws IOException {
+            char symbol = (char) i;
+            buffer.append(symbol);
+            if (symbol == '\n') {
+                sendBuffer();
+                buffer.setLength(0);
+            }
+        }
+
+        private void sendBuffer() {
+            String stringToSent = buffer.toString();
+            for (Iterator<Messenger> it = lifecycleObservers.iterator(); it.hasNext(); ) {
+                Messenger messenger = it.next();
+                try {
+                    Bundle msgData = new Bundle();
+                    msgData.putString(extraId, stringToSent);
+                    Message message = Message.obtain(null, messageId, msgData);
+                    message.setData(msgData);
+                    messenger.send(message);
+                } catch (RemoteException | IllegalStateException e) {
+                    // probably an observer has crashed or terminated, just remove from the set
+                    it.remove();
+                    Log.w(LOG_TAG, "Unable to notify remote node " + Log.getStackTraceString(e));
+                }
+            }
+        }
+    }
+
+    /**
+     * States of Component.
+     */
+    private enum ComponentState {
+        STOPPED,        // stable
+        STARTING,
+        STARTED,
+        RUNNING,        // stable
+        STOPPING,
+    }
+
+    /**
+     * Ensure component is stopped and update state to starting.
+     *
+     * @return true when component state was 'stopped'
+     */
+    boolean ensureStopped() {
+        synchronized (state) {
+            if (state.compareAndSet(ComponentState.STOPPED, ComponentState.STARTING)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Start thread, set 'started' state and wait until thread runs.
+     */
+    void doStart() {
+        // start thread and update state under lock
+        synchronized (state) {
+            workThread.get().start();
+            state.set(ComponentState.STARTED);
+        }
+
+        // waiting until thread is started
+        try {
+            threadStarted.await();
+        } catch (InterruptedException ignored) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)